Add FP32 upcasting for half-precision cross-entropy on large tensors

vishalgoyal316 opened the pull request:
Automatically upcast FP16/BF16 inputs to FP32 when computing cross-entropy loss on large tensors (>10k elements) with reduction, then downcast the result. This improves numerical stability while maintaining performance for smaller tensors.

Fixes #85791

vishalgoyal316 committed:
Add FP32 upcasting for half-precision cross-entropy on large tensors
Automatically upcast FP16/BF16 inputs to FP32 when computing cross-entropy
loss on large tensors (>10k elements) with reduction, then downcast the
result. This improves numerical stability while maintaining performance
for smaller tensors.
Solves: https://github.com/pytorch/pytorch/issues/85791

vadimkantorov commented:
Might it be better to simply always upcast? (and fuse the upcast inside the CE nll_loss function itself to avoid allocation of a temp tensor - can be a huge memory cost for large vocabs)

vishalgoyal316 committed:
Fix fp16/bf16 overflow in CrossEntropyLoss with reduction

Compute in fp32 for half precision types when using reduction to avoid
overflow during accumulation. Leverage log_softmax's fused upcast on
CUDA to avoid materializing temporary fp32 tensors (critical for large
vocabularies).

vishalgoyal316 commented:
> Might it be better to simply always upcast? (and fuse the upcast inside the CE nll_loss function itself to avoid allocation of a temp tensor - can be a huge memory cost for large vocabs)

@vadimkantorov, you are right, I have updated the code based on your feedback, thanks!

vishalgoyal316 commented:
> Why not just do the accumulation in high precision but do the rest of the computation in low precision?

@Skylion007, it is also great idea. I explored it, can I push the code with this approach?

vishalgoyal316 committed:
[loss] Add FP16/BF16 compute with FP32 reduction for cross_entropy

Optimize cross_entropy loss for half precision inputs by computing loss
in native FP16/BF16 and only upcasting to FP32 for the reduction step.

Supports mean/sum reductions with weights and ignore_index handling.

isuruf commented:
Is this issue specific to `nn.CrossEntropyLoss`? Is it there for `nn.NLLLoss` too? If so, I think the issue should be fixed in `nll_loss*` functions.

vishalgoyal316 committed:
Update aten/src/ATen/native/LossNLL.cpp

Co-authored-by: Isuru Fernando <isuruf@gmail.com>

vishalgoyal316 committed:
[loss] Fix FP16/BF16 overflow in NLLLoss with FP32 reduction

Fixes #85791

Implement FP16/BF16 compute with FP32 reduction at the NLLLoss level
instead of CrossEntropyLoss. This architectural choice provides a single
fix for both NLLLoss (direct calls) and CrossEntropyLoss (which uses
NLLLoss internally).

Changes:
- nll_loss_forward_out_cpu_template: Add FP32 reduction for FP16/BF16
- nll_loss_forward_out_cuda_template: Add weight dtype conversion
- cross_entropy_loss_symint: Simplified to delegate to NLLLoss
- Add comprehensive tests for both loss functions

This fixes overflow on large batches while maintaining performance for
the common FP32 case.

vishalgoyal316 commented:
@pytorchbot rebase

isuruf commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -308,33 +308,63 @@ void nll_loss_forward_out_cpu_template(
+
+  // Compute unreduced loss in native precision
+  auto output_unreduced = at::empty_like(input.select(-1, 0));
   AT_DISPATCH_FLOATING_TYPES_AND2(
```
Here, it's the other way. We don't need dispatching to other floating types.

vishalgoyal316 replied:
Sure.

vishalgoyal316 committed:
[NLL Loss] Narrow dispatch sets for CPU NLL loss forward

Split the dispatch logic to avoid unnecessary template instantiations:
- FP32/FP64 fast path now uses AT_DISPATCH_FLOATING_TYPES only
- FP16/BF16 with reduction='none' uses AT_DISPATCH_FLOATING_TYPES_AND2
- FP16/BF16 with reduction uses AT_DISPATCH_REDUCED_FLOATING_TYPES

This reduces binary size and compile time by avoiding:
- Half/BFloat16 instantiations in the FP32/FP64 path
- Float/Double instantiations in the half-precision reduction path

No functional change, just tighter dispatch sets per branch.

vishalgoyal316 commented:
@pytorchbot rebase
